File: modules/lenswire-proxy/ios/WebSocketFrameParser.swift
import Foundation

// Keep in sync with targets/network-packet-tunnel/WebSocketFrameParser.swift

/// RFC 6455 frame parse + display helpers for read-only WebSocket inspect.
enum WebSocketFrames {
  static let maxFrames = 500
  static let maxTextBytes = 64 * 1024
  static let maxBinaryPreview = 256
  static let flushIntervalMs: Int64 = 250
  static let maxPayloadBytes = 16 * 1024 * 1024
  static let maxAssembledBytes = 16 * 1024 * 1024

  static let opcodeContinuation = 0
  static let opcodeText = 1
  static let opcodeBinary = 2
  static let opcodeClose = 8
  static let opcodePing = 9
  static let opcodePong = 10

  static func opcodeName(_ opcode: Int) -> String {
    switch opcode {
    case opcodeContinuation: return "continuation"
    case opcodeText: return "text"
    case opcodeBinary: return "binary"
    case opcodeClose: return "close"
    case opcodePing: return "ping"
    case opcodePong: return "pong"
    default: return "opcode_\(opcode)"
    }
  }

  static func classifyPayload(opcode: Int, payload: Data, rsv: Int = 0) -> [String: Any] {
    if opcode == opcodeClose { return classifyClose(payload) }
    // RSV bits indicate extensions (commonly permessage-deflate) — do not UTF-8 decode as text.
    if rsv != 0 { return classifyBinary(payload, compressed: true) }
    switch opcode {
    case opcodeText, opcodePing, opcodePong:
      return classifyTextish(payload)
    default:
      return classifyBinary(payload)
    }
  }

  private static func classifyClose(_ payload: Data) -> [String: Any] {
    if payload.isEmpty {
      return ["kind": "empty", "size": 0]
    }
    if payload.count >= 2 {
      let code = (Int(payload[0]) << 8) | Int(payload[1])
      let reason: String
      if payload.count > 2 {
        reason = String(data: payload.subdata(in: 2..<payload.count), encoding: .utf8) ?? ""
      } else {
        reason = ""
      }
      let text = reason.isEmpty ? "code=\(code)" : "code=\(code) reason=\(reason)"
      return [
        "kind": "text",
        "text": text,
        "size": payload.count,
        "truncated": false,
        "closeCode": code,
      ]
    }
    return classifyBinary(payload)
  }

  private static func classifyTextish(_ payload: Data) -> [String: Any] {
    if payload.isEmpty {
      return ["kind": "empty", "size": 0]
    }
    var text = String(data: payload, encoding: .utf8) ?? ""
    let size = payload.count
    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
    let kind: String
    if trimmed.first == "{" || trimmed.first == "[" {
      text = prettyJson(text)
      kind = "json"
    } else {
      kind = "text"
    }
    let truncated = payload.count > maxTextBytes || text.utf8.count > maxTextBytes
    let clipped: String
    if text.count > maxTextBytes {
      let idx = text.index(text.startIndex, offsetBy: maxTextBytes)
      clipped = String(text[..<idx]) + "\n\n...truncated..."
    } else {
      clipped = text
    }
    return [
      "kind": kind,
      "text": clipped,
      "size": size,
      "truncated": truncated,
    ]
  }

  private static func prettyJson(_ raw: String) -> String {
    guard let data = raw.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data),
          let pretty = try? JSONSerialization.data(withJSONObject: obj, options: [.prettyPrinted]),
          let text = String(data: pretty, encoding: .utf8)
    else {
      return raw
    }
    return text
  }

  private static func classifyBinary(_ payload: Data, compressed: Bool = false) -> [String: Any] {
    if payload.isEmpty {
      var empty: [String: Any] = ["kind": "empty", "size": 0]
      if compressed { empty["compressed"] = true }
      return empty
    }
    let truncated = payload.count > maxBinaryPreview
    let preview = truncated ? payload.prefix(maxBinaryPreview) : payload
    var out: [String: Any] = [
      "kind": "binary",
      "size": payload.count,
      "truncated": truncated,
      "previewBase64": preview.base64EncodedString(),
    ]
    if compressed { out["compressed"] = true }
    return out
  }

  static func closeCode(from payload: Data) -> Int? {
    guard payload.count >= 2 else { return nil }
    return (Int(payload[0]) << 8) | Int(payload[1])
  }

  static func displayFrame(dir: String, opcode: Int, payload: Data, rsv: Int = 0, atMs: Int64? = nil) -> [String: Any] {
    let at = atMs ?? Int64(Date().timeIntervalSince1970 * 1000)
    return [
      "id": UUID().uuidString,
      "at": at,
      "dir": dir,
      "opcode": opcodeName(opcode),
      "size": payload.count,
      "payload": classifyPayload(opcode: opcode, payload: payload, rsv: rsv),
    ]
  }
}

/// Incremental RFC 6455 frame parser. Forwards bytes separately; this only inspects.
final class WebSocketFrameParser {
  struct RawFrame {
    let fin: Bool
    let opcode: Int
    let payload: Data
    let rsv: Int
    let oversized: Bool

    init(fin: Bool, opcode: Int, payload: Data, rsv: Int = 0, oversized: Bool = false) {
      self.fin = fin
      self.opcode = opcode
      self.payload = payload
      self.rsv = rsv
      self.oversized = oversized
    }
  }

  private var pending = Data()

  func append(_ bytes: Data) {
    guard !bytes.isEmpty else { return }
    pending.append(bytes)
  }

  func drain() -> [RawFrame] {
    var index = 0
    var out: [RawFrame] = []
    while true {
      guard let parsed = parseOne(pending, start: index) else { break }
      out.append(parsed.frame)
      index = parsed.nextIndex
    }
    if index > 0 {
      if index < pending.count {
        pending = Data(pending[index...])
      } else {
        pending.removeAll(keepingCapacity: true)
      }
    }
    return out
  }

  private struct ParseResult {
    let frame: RawFrame
    let nextIndex: Int
  }

  private func parseOne(_ buf: Data, start: Int) -> ParseResult? {
    guard buf.count - start >= 2 else { return nil }
    let b0 = Int(buf[start])
    let b1 = Int(buf[start + 1])
    let fin = (b0 & 0x80) != 0
    let rsv = (b0 & 0x70) >> 4
    let opcode = b0 & 0x0f
    let masked = (b1 & 0x80) != 0
    var payloadLen = Int(b1 & 0x7f)
    var headerLen = 2

    if payloadLen == 126 {
      guard buf.count - start >= 4 else { return nil }
      payloadLen = (Int(buf[start + 2]) << 8) | Int(buf[start + 3])
      headerLen = 4
    } else if payloadLen == 127 {
      guard buf.count - start >= 10 else { return nil }
      var value: UInt64 = 0
      for i in 0..<8 {
        value = (value << 8) | UInt64(buf[start + 2 + i])
      }
      if value > UInt64(Int.max) { return nil }
      payloadLen = Int(value)
      headerLen = 10
    }

    if payloadLen > WebSocketFrames.maxPayloadBytes {
      let maskLen = masked ? 4 : 0
      let total = headerLen + maskLen + payloadLen
      guard buf.count - start >= total else { return nil }
      return ParseResult(
        frame: RawFrame(fin: fin, opcode: opcode, payload: Data(), rsv: rsv, oversized: true),
        nextIndex: start + total
      )
    }

    let maskLen = masked ? 4 : 0
    let total = headerLen + maskLen + payloadLen
    guard buf.count - start >= total else { return nil }

    let maskStart = start + headerLen
    let payloadStart = maskStart + maskLen
    var payload = Data(buf[payloadStart..<(payloadStart + payloadLen)])
    if masked {
      let k0 = buf[maskStart]
      let k1 = buf[maskStart + 1]
      let k2 = buf[maskStart + 2]
      let k3 = buf[maskStart + 3]
      for i in 0..<payload.count {
        let key: UInt8
        switch i % 4 {
        case 0: key = k0
        case 1: key = k1
        case 2: key = k2
        default: key = k3
        }
        payload[i] ^= key
      }
    }
    return ParseResult(
      frame: RawFrame(fin: fin, opcode: opcode, payload: payload, rsv: rsv),
      nextIndex: start + total
    )
  }
}

/// Assembles fragmented data messages into one display unit. Control frames emit immediately.
final class WebSocketMessageAssembler {
  struct Assembled {
    let opcode: Int
    let payload: Data
    let rsv: Int
    let omitted: Bool

    init(opcode: Int, payload: Data, rsv: Int = 0, omitted: Bool = false) {
      self.opcode = opcode
      self.payload = payload
      self.rsv = rsv
      self.omitted = omitted
    }
  }

  private var startedOpcode: Int?
  private var startedRsv: Int = 0
  private var chunks = Data()

  func accept(_ frame: WebSocketFrameParser.RawFrame) -> Assembled? {
    let opcode = frame.opcode

    if frame.oversized {
      reset()
      return frame.fin ? Assembled(opcode: opcode, payload: Data(), omitted: true) : nil
    }

    if opcode == WebSocketFrames.opcodeClose
      || opcode == WebSocketFrames.opcodePing
      || opcode == WebSocketFrames.opcodePong
    {
      return Assembled(opcode: opcode, payload: frame.payload, rsv: frame.rsv)
    }

    if opcode == WebSocketFrames.opcodeText || opcode == WebSocketFrames.opcodeBinary {
      startedOpcode = opcode
      startedRsv = frame.rsv
      chunks = Data()
      if !appendChunk(frame.payload) {
        reset()
        return frame.fin ? Assembled(opcode: opcode, payload: Data(), omitted: true) : nil
      }
      if frame.fin {
        let out = Assembled(opcode: startedOpcode!, payload: chunks, rsv: startedRsv)
        reset()
        return out
      }
      return nil
    }

    if opcode == WebSocketFrames.opcodeContinuation {
      guard let started = startedOpcode else {
        return frame.fin ? Assembled(opcode: WebSocketFrames.opcodeBinary, payload: frame.payload, rsv: frame.rsv) : nil
      }
      if !appendChunk(frame.payload) {
        reset()
        return frame.fin ? Assembled(opcode: started, payload: Data(), omitted: true) : nil
      }
      if frame.fin {
        let out = Assembled(opcode: started, payload: chunks, rsv: startedRsv)
        reset()
        return out
      }
      return nil
    }

    return frame.fin ? Assembled(opcode: opcode, payload: frame.payload, rsv: frame.rsv) : nil
  }

  private func appendChunk(_ payload: Data) -> Bool {
    if chunks.count + payload.count > WebSocketFrames.maxAssembledBytes {
      return false
    }
    chunks.append(payload)
    return true
  }

  private func reset() {
    startedOpcode = nil
    startedRsv = 0
    chunks = Data()
  }
}

/// Throttled writer that grows `wsFrames` on an existing capture.
final class WsFrameCaptureRecorder {
  private let captureId: String
  private let lock = NSLock()
  private var pending: [[String: Any]] = []
  private var lastFlushMs: Int64 = 0
  private var totalRecorded = 0
  private var omitted = false
  private var compressed = false
  private var idleFlushWorkItem: DispatchWorkItem?
  private var recorderDone = false
  private var sessionClosed = false
  private var endReason: String?
  private var endCloseCode: Int?
  private var endedAtMs: Int64?

  init(captureId: String) {
    self.captureId = captureId
  }

  func record(dir: String, opcode: Int, payload: Data, rsv: Int = 0) {
    lock.lock()
    defer { lock.unlock() }
    guard !recorderDone else { return }
    if rsv != 0 { compressed = true }
    let isClose = opcode == WebSocketFrames.opcodeClose
    if totalRecorded >= WebSocketFrames.maxFrames && !isClose {
      if !omitted {
        omitted = true
        flushLocked(force: true)
      }
      return
    }
    totalRecorded += 1
    pending.append(WebSocketFrames.displayFrame(dir: dir, opcode: opcode, payload: payload, rsv: rsv))
    if isClose {
      markClosedLocked(reason: "close_frame", closeCode: WebSocketFrames.closeCode(from: payload))
    }
    flushLocked(force: isClose)
    if !isClose { scheduleIdleFlushLocked() }
  }

  func markOmitted() {
    lock.lock()
    defer { lock.unlock() }
    guard !recorderDone else { return }
    if !omitted {
      omitted = true
      flushLocked(force: true)
    }
  }

  func markCompressed() {
    lock.lock()
    defer { lock.unlock() }
    guard !recorderDone else { return }
    compressed = true
  }

  func markClosed(reason: String, closeCode: Int? = nil) {
    lock.lock()
    defer { lock.unlock() }
    markClosedLocked(reason: reason, closeCode: closeCode)
    flushLocked(force: true)
    idleFlushWorkItem?.cancel()
    idleFlushWorkItem = nil
  }

  func flush(force: Bool = true) {
    lock.lock()
    defer { lock.unlock() }
    flushLocked(force: force)
    if force {
      idleFlushWorkItem?.cancel()
      idleFlushWorkItem = nil
    }
  }

  private func markClosedLocked(reason: String, closeCode: Int?) {
    if sessionClosed {
      let existing = endReason
      if existing == "close_frame" || reason != "close_frame" { return }
    }
    sessionClosed = true
    endReason = reason
    if let closeCode { endCloseCode = closeCode }
    if endedAtMs == nil {
      endedAtMs = Int64(Date().timeIntervalSince1970 * 1000)
    }
  }

  private func scheduleIdleFlushLocked() {
    idleFlushWorkItem?.cancel()
    let item = DispatchWorkItem { [weak self] in
      self?.flush(force: true)
    }
    idleFlushWorkItem = item
    DispatchQueue.global(qos: .utility).asyncAfter(
      deadline: .now() + .milliseconds(Int(WebSocketFrames.flushIntervalMs)),
      execute: item
    )
  }

  /// Snapshot pending frames and persist off the relay path.
  private func flushLocked(force: Bool) {
    let now = Int64(Date().timeIntervalSince1970 * 1000)
    if !force && pending.isEmpty && !sessionClosed { return }
    if !force
      && now - lastFlushMs < WebSocketFrames.flushIntervalMs
      && pending.count < 8
    {
      return
    }
    let shouldFlushClosed = force && sessionClosed
    if pending.isEmpty && !(force && (omitted || compressed || shouldFlushClosed)) { return }
    let batch = pending
    pending.removeAll(keepingCapacity: true)
    lastFlushMs = now
    let closed = sessionClosed
    let reason = endReason
    let closeCode = endCloseCode
    let endedAt = endedAtMs
    let omit = omitted
    let comp = compressed
    let id = captureId
    Self.flushQueue.async {
      if batch.isEmpty && closed {
        LenswireShared.markWsClosed(
          id: id,
          reason: reason ?? "eof",
          closeCode: closeCode,
          endedAt: endedAt ?? now
        )
      } else {
        LenswireShared.appendWsFrames(
          id: id,
          frames: batch,
          omitted: omit,
          compressed: comp,
          wsClosed: closed,
          endedAt: closed ? endedAt : nil,
          wsEndReason: closed ? reason : nil,
          wsCloseCode: closed ? closeCode : nil
        )
      }
    }
  }

  private static let flushQueue = DispatchQueue(label: "lenswire.ws.flush", qos: .utility)
}

File: modules/lenswire-proxy/android/src/main/java/expo/modules/lenswireproxy/WebSocketFrameParser.kt
package expo.modules.lenswireproxy

import android.util.Base64
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/** RFC 6455 frame parse + display helpers for read-only WebSocket inspect. */
internal object WebSocketFrames {
  const val MAX_FRAMES = 500
  const val MAX_TEXT_BYTES = 64 * 1024
  const val MAX_BINARY_PREVIEW = 256
  const val FLUSH_INTERVAL_MS = 250L
  const val MAX_PAYLOAD_BYTES = 16 * 1024 * 1024
  const val MAX_ASSEMBLED_BYTES = 16 * 1024 * 1024

  const val OPCODE_CONTINUATION = 0
  const val OPCODE_TEXT = 1
  const val OPCODE_BINARY = 2
  const val OPCODE_CLOSE = 8
  const val OPCODE_PING = 9
  const val OPCODE_PONG = 10

  fun opcodeName(opcode: Int): String = when (opcode) {
    OPCODE_CONTINUATION -> "continuation"
    OPCODE_TEXT -> "text"
    OPCODE_BINARY -> "binary"
    OPCODE_CLOSE -> "close"
    OPCODE_PING -> "ping"
    OPCODE_PONG -> "pong"
    else -> "opcode_$opcode"
  }

  fun classifyPayload(opcode: Int, payload: ByteArray, rsv: Int = 0): Map<String, Any?> {
    if (opcode == OPCODE_CLOSE) return classifyClose(payload)
    // RSV bits indicate extensions (commonly permessage-deflate) — do not UTF-8 decode as text.
    if (rsv != 0) return classifyBinary(payload, compressed = true)
    return when (opcode) {
      OPCODE_TEXT, OPCODE_PING, OPCODE_PONG -> classifyTextish(payload)
      OPCODE_BINARY -> classifyBinary(payload)
      else -> classifyBinary(payload)
    }
  }

  private fun classifyClose(payload: ByteArray): Map<String, Any?> {
    if (payload.isEmpty()) {
      return mapOf("kind" to "empty", "size" to 0)
    }
    if (payload.size >= 2) {
      val code = ((payload[0].toInt() and 0xff) shl 8) or (payload[1].toInt() and 0xff)
      val reason = if (payload.size > 2) {
        String(payload, 2, payload.size - 2, Charsets.UTF_8)
      } else {
        ""
      }
      val text = if (reason.isEmpty()) "code=$code" else "code=$code reason=$reason"
      return mapOf(
        "kind" to "text",
        "text" to text,
        "size" to payload.size,
        "truncated" to false,
        "closeCode" to code,
      )
    }
    return classifyBinary(payload)
  }

  private fun classifyTextish(payload: ByteArray): Map<String, Any?> {
    if (payload.isEmpty()) {
      return mapOf("kind" to "empty", "size" to 0)
    }
    var text = String(payload, Charsets.UTF_8)
    val size = payload.size
    val trimmed = text.trimStart()
    val kind = if (trimmed.firstOrNull() in setOf('{', '[')) {
      text = CaptureFormatting.prettyJson(text)
      "json"
    } else {
      "text"
    }
    val truncated = payload.size > MAX_TEXT_BYTES || text.toByteArray(Charsets.UTF_8).size > MAX_TEXT_BYTES
    val clipped = if (text.length > MAX_TEXT_BYTES) {
      text.substring(0, MAX_TEXT_BYTES) + "\n\n...truncated..."
    } else {
      text
    }
    return mapOf(
      "kind" to kind,
      "text" to clipped,
      "size" to size,
      "truncated" to truncated,
    )
  }

  private fun classifyBinary(payload: ByteArray, compressed: Boolean = false): Map<String, Any?> {
    if (payload.isEmpty()) {
      return buildMap {
        put("kind", "empty")
        put("size", 0)
        if (compressed) put("compressed", true)
      }
    }
    val truncated = payload.size > MAX_BINARY_PREVIEW
    val preview = if (truncated) payload.copyOf(MAX_BINARY_PREVIEW) else payload
    return buildMap {
      put("kind", "binary")
      put("size", payload.size)
      put("truncated", truncated)
      put("previewBase64", Base64.encodeToString(preview, Base64.NO_WRAP))
      if (compressed) put("compressed", true)
    }
  }

  fun closeCodeFromPayload(payload: ByteArray): Int? {
    if (payload.size < 2) return null
    return ((payload[0].toInt() and 0xff) shl 8) or (payload[1].toInt() and 0xff)
  }

  fun displayFrame(
    dir: String,
    opcode: Int,
    payload: ByteArray,
    rsv: Int = 0,
    atMs: Long = System.currentTimeMillis(),
  ): Map<String, Any?> = mapOf(
    "id" to UUID.randomUUID().toString(),
    "at" to atMs,
    "dir" to dir,
    "opcode" to opcodeName(opcode),
    "size" to payload.size,
    "payload" to classifyPayload(opcode, payload, rsv),
  )
}

/** Incremental RFC 6455 frame parser. Forwards bytes separately; this only inspects. */
internal class WebSocketFrameParser {
  private val pending = ByteArrayOutputStream()

  data class RawFrame(
    val fin: Boolean,
    val opcode: Int,
    /** Unmasked payload for display (empty when [oversized]). */
    val payload: ByteArray,
    val rsv: Int = 0,
    /** True when payload exceeded MAX_PAYLOAD_BYTES and was skipped. */
    val oversized: Boolean = false,
  )

  fun append(bytes: ByteArray, offset: Int = 0, length: Int = bytes.size) {
    if (length <= 0) return
    pending.write(bytes, offset, length)
  }

  fun append(bytes: ByteArray) = append(bytes, 0, bytes.size)

  fun drain(): List<RawFrame> {
    val buf = pending.toByteArray()
    var index = 0
    val out = ArrayList<RawFrame>()
    while (true) {
      val parsed = parseOne(buf, index) ?: break
      out.add(parsed.frame)
      index = parsed.nextIndex
    }
    if (index > 0) {
      pending.reset()
      if (index < buf.size) {
        pending.write(buf, index, buf.size - index)
      }
    }
    return out
  }

  private data class ParseResult(val frame: RawFrame, val nextIndex: Int)

  private fun parseOne(buf: ByteArray, start: Int): ParseResult? {
    if (buf.size - start < 2) return null
    val b0 = buf[start].toInt() and 0xff
    val b1 = buf[start + 1].toInt() and 0xff
    val fin = (b0 and 0x80) != 0
    val rsv = (b0 and 0x70) shr 4
    val opcode = b0 and 0x0f
    val masked = (b1 and 0x80) != 0
    var payloadLen = (b1 and 0x7f).toLong()
    var headerLen = 2

    when (payloadLen) {
      126L -> {
        if (buf.size - start < 4) return null
        payloadLen = ((buf[start + 2].toInt() and 0xff) shl 8 or (buf[start + 3].toInt() and 0xff)).toLong()
        headerLen = 4
      }
      127L -> {
        if (buf.size - start < 10) return null
        val bb = ByteBuffer.wrap(buf, start + 2, 8).order(ByteOrder.BIG_ENDIAN)
        payloadLen = bb.long
        if (payloadLen < 0) return null
        headerLen = 10
      }
    }

    if (payloadLen > WebSocketFrames.MAX_PAYLOAD_BYTES) {
      val maskLen = if (masked) 4 else 0
      val total = headerLen + maskLen + payloadLen
      if (buf.size - start < total) return null
      return ParseResult(
        frame = RawFrame(
          fin = fin,
          opcode = opcode,
          payload = ByteArray(0),
          rsv = rsv,
          oversized = true,
        ),
        nextIndex = start + total.toInt(),
      )
    }

    val maskLen = if (masked) 4 else 0
    val total = headerLen + maskLen + payloadLen.toInt()
    if (buf.size - start < total) return null

    val maskStart = start + headerLen
    val payloadStart = maskStart + maskLen
    val payload = buf.copyOfRange(payloadStart, payloadStart + payloadLen.toInt())
    if (masked) {
      val key0 = buf[maskStart].toInt()
      val key1 = buf[maskStart + 1].toInt()
      val key2 = buf[maskStart + 2].toInt()
      val key3 = buf[maskStart + 3].toInt()
      for (i in payload.indices) {
        val key = when (i % 4) {
          0 -> key0
          1 -> key1
          2 -> key2
          else -> key3
        }
        payload[i] = (payload[i].toInt() xor key).toByte()
      }
    }
    return ParseResult(
      frame = RawFrame(fin = fin, opcode = opcode, payload = payload, rsv = rsv),
      nextIndex = start + total,
    )
  }
}

/**
 * Assembles fragmented data messages (opcode text/binary + continuation) into one display unit.
 * Control frames are emitted immediately.
 */
internal class WebSocketMessageAssembler {
  data class Assembled(
    val opcode: Int,
    val payload: ByteArray,
    val rsv: Int = 0,
    /** True when frame/message was dropped (oversized or assemble cap); do not display. */
    val omitted: Boolean = false,
  )

  private var startedOpcode: Int? = null
  private var startedRsv: Int = 0
  private val chunks = ByteArrayOutputStream()

  /** Returns a completed message, or null if still fragmenting. */
  fun accept(frame: WebSocketFrameParser.RawFrame): Assembled? {
    val opcode = frame.opcode

    if (frame.oversized) {
      // Drop in-progress assemble; never emit empty placeholder messages.
      reset()
      return if (frame.fin) Assembled(opcode = opcode, payload = ByteArray(0), omitted = true) else null
    }

    if (opcode == WebSocketFrames.OPCODE_CLOSE ||
      opcode == WebSocketFrames.OPCODE_PING ||
      opcode == WebSocketFrames.OPCODE_PONG
    ) {
      return Assembled(opcode = opcode, payload = frame.payload, rsv = frame.rsv)
    }

    if (opcode == WebSocketFrames.OPCODE_TEXT || opcode == WebSocketFrames.OPCODE_BINARY) {
      startedOpcode = opcode
      startedRsv = frame.rsv
      chunks.reset()
      if (!appendChunk(frame.payload)) {
        reset()
        return if (frame.fin) Assembled(opcode = opcode, payload = ByteArray(0), omitted = true) else null
      }
      if (frame.fin) {
        val out = Assembled(startedOpcode!!, chunks.toByteArray(), startedRsv)
        reset()
        return out
      }
      return null
    }

    if (opcode == WebSocketFrames.OPCODE_CONTINUATION) {
      if (startedOpcode == null) {
        return if (frame.fin) {
          Assembled(WebSocketFrames.OPCODE_BINARY, frame.payload, frame.rsv)
        } else {
          null
        }
      }
      if (!appendChunk(frame.payload)) {
        val op = startedOpcode!!
        reset()
        return if (frame.fin) Assembled(opcode = op, payload = ByteArray(0), omitted = true) else null
      }
      if (frame.fin) {
        val out = Assembled(startedOpcode!!, chunks.toByteArray(), startedRsv)
        reset()
        return out
      }
      return null
    }

    return if (frame.fin) Assembled(opcode, frame.payload, frame.rsv) else null
  }

  private fun appendChunk(payload: ByteArray): Boolean {
    if (chunks.size().toLong() + payload.size > WebSocketFrames.MAX_ASSEMBLED_BYTES) {
      return false
    }
    chunks.write(payload)
    return true
  }

  private fun reset() {
    startedOpcode = null
    startedRsv = 0
    chunks.reset()
  }
}

/** Throttled writer that grows `wsFrames` on an existing capture. */
internal class WsFrameCaptureRecorder(
  private val context: android.content.Context,
  private val captureId: String,
) {
  private val lock = Any()
  private val pending = ArrayList<Map<String, Any?>>()
  private var lastFlushMs = 0L
  private var totalRecorded = 0
  private var omitted = false
  private var compressed = false
  private var idleFlushFuture: ScheduledFuture<*>? = null
  private val recorderDone = AtomicBoolean(false)
  private val sessionClosed = AtomicBoolean(false)
  private var endReason: String? = null
  private var endCloseCode: Int? = null
  private var endedAtMs: Long? = null

  fun record(dir: String, opcode: Int, payload: ByteArray, rsv: Int = 0) {
    synchronized(lock) {
      if (recorderDone.get()) return
      if (rsv != 0) compressed = true
      val isClose = opcode == WebSocketFrames.OPCODE_CLOSE
      if (totalRecorded >= WebSocketFrames.MAX_FRAMES && !isClose) {
        if (!omitted) {
          omitted = true
          scheduleFlushLocked(force = true)
        }
        return
      }
      totalRecorded += 1
      pending.add(WebSocketFrames.displayFrame(dir, opcode, payload, rsv))
      if (isClose) {
        val code = WebSocketFrames.closeCodeFromPayload(payload)
        markClosedLocked("close_frame", code)
      }
      scheduleFlushLocked(force = isClose)
      if (!isClose) scheduleIdleFlushLocked()
    }
  }

  fun markOmitted() {
    synchronized(lock) {
      if (recorderDone.get()) return
      if (!omitted) {
        omitted = true
        scheduleFlushLocked(force = true)
      }
    }
  }

  fun markCompressed() {
    synchronized(lock) {
      if (recorderDone.get()) return
      compressed = true
    }
  }

  /** Once: persist wsClosed on the capture. Prefer close_frame over eof/error. */
  fun markClosed(reason: String, closeCode: Int? = null) {
    synchronized(lock) {
      markClosedLocked(reason, closeCode)
      scheduleFlushLocked(force = true)
      idleFlushFuture?.cancel(false)
      idleFlushFuture = null
    }
  }

  fun flush(force: Boolean = true) {
    synchronized(lock) {
      scheduleFlushLocked(force = force)
      if (force) {
        idleFlushFuture?.cancel(false)
        idleFlushFuture = null
      }
    }
  }

  fun close() {
    markClosed("eof")
    recorderDone.set(true)
    flush(force = true)
    synchronized(lock) {
      idleFlushFuture?.cancel(false)
      idleFlushFuture = null
    }
  }

  private fun markClosedLocked(reason: String, closeCode: Int?) {
    if (sessionClosed.get()) {
      val existing = endReason
      if (existing == "close_frame" || reason != "close_frame") return
    }
    sessionClosed.set(true)
    endReason = reason
    if (closeCode != null) endCloseCode = closeCode
    if (endedAtMs == null) endedAtMs = System.currentTimeMillis()
  }

  private fun scheduleIdleFlushLocked() {
    idleFlushFuture?.cancel(false)
    idleFlushFuture = idleFlushScheduler.schedule({
      if (recorderDone.get()) return@schedule
      flush(force = true)
    }, WebSocketFrames.FLUSH_INTERVAL_MS, TimeUnit.MILLISECONDS)
  }

  /**
   * Snapshot pending frames and persist off the relay thread.
   * CaptureStore I/O must never block WebSocket forwarding (reconnect latency).
   */
  private fun scheduleFlushLocked(force: Boolean) {
    val now = System.currentTimeMillis()
    if (!force && pending.isEmpty() && !sessionClosed.get()) return
    if (!force && now - lastFlushMs < WebSocketFrames.FLUSH_INTERVAL_MS && pending.size < 8) {
      return
    }
    val shouldFlushClosed = force && sessionClosed.get()
    if (pending.isEmpty() && !(force && (omitted || compressed || shouldFlushClosed))) return
    val batch = ArrayList(pending)
    pending.clear()
    lastFlushMs = now
    val closed = sessionClosed.get()
    val reason = endReason
    val closeCode = endCloseCode
    val endedAt = endedAtMs
    val omit = omitted
    val comp = compressed
    val id = captureId
    val ctx = context
    idleFlushScheduler.execute {
      runCatching {
        if (batch.isEmpty() && closed) {
          CaptureStore.markWsClosed(
            ctx,
            id,
            reason ?: "eof",
            closeCode,
            endedAt ?: System.currentTimeMillis(),
          )
        } else {
          CaptureStore.appendWsFrames(
            ctx,
            id,
            batch,
            omit,
            comp,
            wsClosed = closed,
            endedAt = if (closed) endedAt else null,
            wsEndReason = if (closed) reason else null,
            wsCloseCode = if (closed) closeCode else null,
          )
        }
      }
    }
  }

  companion object {
    private val idleFlushScheduler = Executors.newSingleThreadScheduledExecutor { r ->
      Thread(r, "lenswire-ws-flush").apply { isDaemon = true }
    }
  }
}
